Adding AAD authentication client & sample, updating README.md file

rangothedog opened the pull request:
Adding AAD authentication client & sample, updating README.md file.

Commit:
Adding AAD authentication client & sample, updating README.md file

Commit:
Adding AAD authentication client & sample, updating README.md file

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.AI.Translation.Text](https://apiview.dev/Assemblies/Review/f552c1fa031940a5aee1cee53bc90fec)

JoshLove-msft committed:
Split pipelines (#40586)

* Split pipelines

* fix paths

* Use absolute paths

azure-sdk committed:
Increment version for storage releases (#40594)

jsquire committed:
[Event Hubs] Processor fixes (#40593)

* [Event Hubs] Processor fixes

The focus of these changes is to address a race condition where calling
`IsRunning` concurrently with starting/stopping the processor could cause a
semaphore to be double-released.  As part of this fix, the approach used to
detect when a semaphore should be released has been changed to consider the
semaphore count rather than using a sentinal variable.

Also fixed was an issue where the processor would leak a quota exceeded
exception during startup validation if there were more than 4 other processors
starting concurrently.  The partition chosen for validation is now randomized and
a quota exceeded response is now considered a successful validation.

* Update sdk/eventhub/Azure.Messaging.EventHubs.Processor/CHANGELOG.md

Co-authored-by: Madalyn Redding <66138537+m-redding@users.noreply.github.com>

* Update sdk/eventhub/Azure.Messaging.EventHubs/CHANGELOG.md

Co-authored-by: Madalyn Redding <66138537+m-redding@users.noreply.github.com>

---------

Co-authored-by: Madalyn Redding <66138537+m-redding@users.noreply.github.com>

Commit:
Merge branch 'rangothedog/AADAuthSample' of https://github.com/Azure/azure-sdk-for-net into rangothedog/AADAuthSample

scottaddie commented on sdk/translation/Azure.AI.Translation.Text/README.md:
```diff
@@ -65,6 +65,26 @@ string region = "<Text Translator Azure Region>";
 
+#### Create `TextTranslationClient` with Entra Credential
+
+Client API key authentication is used in most of the examples, but you can also authenticate with Entra (Azure Active Directory) using the [Azure Identity library][azure_identity].  To use the [DefaultAzureCredential][DefaultAzureCredential] provider shown below, please install the Azure.Identity package:
```
```suggestion
Client API key authentication is used in most of the examples, but you can also authenticate with Entra ID (Azure Active Directory) using the [Azure Identity library][azure_identity].  To use the [DefaultAzureCredential][DefaultAzureCredential] provider shown below, please install the Azure.Identity package:
```

scottaddie commented on sdk/translation/Azure.AI.Translation.Text/README.md:
````diff
@@ -65,6 +65,26 @@ string region = "<Text Translator Azure Region>";
+dotnet add package Azure.Identity
+```
+
+Create a [custom subdomain][custom_subdomain] for your resource in order to use this type of authentication.  Use this value for the `endpoint` variable for `Text Translator Custom Endpoint`.  Only custom subdomains are supported by the SDK.  Outside of the SDK you can use AAD authorization for Machine Tranlsation endpoints as detailed [here][mt_endpoints].
````
```suggestion
Create a [custom subdomain][custom_subdomain] for your resource in order to use this type of authentication.  Use this value for the `endpoint` variable for `Text Translator Custom Endpoint`.  Only custom subdomains are supported by the SDK.  Outside of the SDK, you can use Entra authorization for Machine Translation endpoints, as detailed [here][mt_endpoints].
```

scottaddie commented on sdk/translation/Azure.AI.Translation.Text/README.md:
```diff
@@ -65,6 +65,26 @@ string region = "<Text Translator Azure Region>";
+
+Create a [custom subdomain][custom_subdomain] for your resource in order to use this type of authentication.  Use this value for the `endpoint` variable for `Text Translator Custom Endpoint`.  Only custom subdomains are supported by the SDK.  Outside of the SDK you can use AAD authorization for Machine Tranlsation endpoints as detailed [here][mt_endpoints].
+
+You will also need to [register a new AAD application][register_aad_app] and [grant access][aad_grant_access] to your Translator resource by assigning the `"Cognitive Services User"` role to your service principal.  Additional information about Entra authentication is available [here][custom_details].
```
```suggestion
You will also need to [register a new Entra application][register_aad_app] and [grant access][aad_grant_access] to your Translator resource by assigning the `"Cognitive Services User"` role to your service principal.  Additional information about Entra authentication is available [here][custom_details].
```

scottaddie commented on sdk/translation/Azure.AI.Translation.Text/README.md:
```diff
@@ -65,6 +65,26 @@ string region = "<Text Translator Azure Region>";
+
+You will also need to [register a new AAD application][register_aad_app] and [grant access][aad_grant_access] to your Translator resource by assigning the `"Cognitive Services User"` role to your service principal.  Additional information about Entra authentication is available [here][custom_details].
+
+Set the values of the `client ID`, `tenant ID`, and `client secret` of the AAD application as environment variables: `AZURE_CLIENT_ID`, `AZURE_TENANT_ID`, `AZURE_CLIENT_SECRET`.  The DefaultAzureCredential constructor uses these variables to create your credentials.
```
```suggestion
Set the values of the `client ID`, `tenant ID`, and `client secret` of the Entra application as environment variables: `AZURE_CLIENT_ID`, `AZURE_TENANT_ID`, `AZURE_CLIENT_SECRET`.  The DefaultAzureCredential constructor uses these variables to create your credentials.
```

rangothedog committed:
Update sdk/translation/Azure.AI.Translation.Text/README.md

Changing AAD to Entra.

Co-authored-by: Scott Addie <10702007+scottaddie@users.noreply.github.com>

rangothedog committed:
Update sdk/translation/Azure.AI.Translation.Text/README.md

Add ID to README.md.

Co-authored-by: Scott Addie <10702007+scottaddie@users.noreply.github.com>

rangothedog committed:
Update sdk/translation/Azure.AI.Translation.Text/README.md

Fix spelling error on Translation.

Co-authored-by: Scott Addie <10702007+scottaddie@users.noreply.github.com>

rangothedog committed:
Update sdk/translation/Azure.AI.Translation.Text/README.md

Change AAD to Entra.

Co-authored-by: Scott Addie <10702007+scottaddie@users.noreply.github.com>

scottaddie commented on sdk/translation/Azure.AI.Translation.Text/README.md:
```diff
@@ -65,6 +65,26 @@ string region = "<Text Translator Azure Region>";
 
+#### Create `TextTranslationClient` with Entra Credential
+
+Client API key authentication is used in most of the examples, but you can also authenticate with Entra ID (Azure Active Directory) using the [Azure Identity library][azure_identity].  To use the [DefaultAzureCredential][DefaultAzureCredential] provider shown below, please install the Azure.Identity package:
```
```suggestion
Client API key authentication is used in most of the examples, but you can also authenticate with Microsoft Entra ID using the [Azure Identity library][azure_identity]. To use the [DefaultAzureCredential][DefaultAzureCredential] provider shown below, install the Azure.Identity package:
```

scottaddie commented on sdk/translation/Azure.AI.Translation.Text/README.md:
```diff
@@ -449,7 +469,12 @@ This project has adopted the [Microsoft Open Source Code of Conduct][code_of_con
-
+[azure_identity]: https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/identity/Azure.Identity/README.md
+[DefaultAzureCredential]: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/README.md#defaultazurecredential
+[register_aad_app]: https://docs.microsoft.com/azure/cognitive-services/authentication#assign-a-role-to-a-service-principal
+[aad_grant_access]: https://docs.microsoft.com/azure/cognitive-services/authentication#assign-a-role-to-a-service-principal
+[custom_subdomain]: https://docs.microsoft.com/azure/cognitive-services/authentication#create-a-resource-with-a-custom-subdomain
```
```suggestion
[register_aad_app]: https://learn.microsoft.com/azure/cognitive-services/authentication#assign-a-role-to-a-service-principal
[aad_grant_access]: https://learn.microsoft.com/azure/cognitive-services/authentication#assign-a-role-to-a-service-principal
[custom_subdomain]: https://learn.microsoft.com/azure/cognitive-services/authentication#create-a-resource-with-a-custom-subdomain
```

scottaddie commented on sdk/translation/Azure.AI.Translation.Text/src/Custom/TextTranslationClient.cs:
```diff
@@ -118,6 +119,16 @@ public TextTranslationClient(TokenCredential credential, Uri endpoint, TextTrans
+        /// <param name="credential">DefaultAzureCredential</param>
+        /// <param name="endpoint">Service Endpoint</param>
+        /// <param name="options">Translate Client Options</param>
+        public TextTranslationClient(DefaultAzureCredential credential, Uri endpoint, TextTranslationClientOptions options = default) : this(credential as TokenCredential, endpoint, options)
```
I'm not sure I understand the purpose of this code. What does this ctor allow the customer to do that they can't already do with the following ctor? `DefaultAzureCredential` is a `TokenCredential`.

```csharp
public TextTranslationClient(TokenCredential credential, Uri endpoint, TextTranslationClientOptions options = default) : this(endpoint, options)
```

jsquire replied:
This won't pass architect review, for the reasons that Scott is calling out.   You'll want to remove this.

//cc: @tg-msft, @KrzysztofCwalina

jsquire reviewed (changes requested):
Please back out changes to Event Hubs, Service Bus, and Storage.  They should not be part of this PR.

rangothedog committed:
Update sdk/translation/Azure.AI.Translation.Text/README.md

Adding Microsoft to 'Microsoft Entra'.

Co-authored-by: Scott Addie <10702007+scottaddie@users.noreply.github.com>

jsquire commented on sdk/translation/Azure.AI.Translation.Text/tests/Samples/SampleSnippets.cs:
```diff
@@ -127,6 +127,24 @@ public TextTranslationClient CreateTextTranslationClientWithToken()
+            string endpoint = TestEnvironment.CustomEndpoint;
+#endif
+            DefaultAzureCredential credential = new DefaultAzureCredential();
+            TextTranslationClient client = new TextTranslationClient(credential, new Uri(endpoint));
```
This will use the existing `TokenCredential` constructor just fine.  There's no need for a specific overload.

rangothedog committed:
Update sdk/translation/Azure.AI.Translation.Text/README.md

Change urls docs -> learn.

Co-authored-by: Scott Addie <10702007+scottaddie@users.noreply.github.com>
